aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,639 @@
+    const PROPOSAL_RESOLVED_NOT_PASSED: u8 = 2;
+
+    const DEFALT_VOTING_WEIGHT: u64 = 1;
+
+    struct DAO has key {
+        /// Name of the DAO
+        name: String,
+        /// The threshold that the proposal can resolve. For example, 50% means at least 50% of NFT holder should vote to resolve the proposal
+        resolve_threshold: ResolveThreshold,
+        /// The NFT Collection that is used to govern the DAO
+        governance_token: GovernanceToken, // This is the governance token or NFT used for voting
+        /// The address of the creator account this DAO owns. THis account is a resource account that this DAO is an adminstrator of the controlled account
+        controlled_account: address, // This is the nft creator account that this DAO governs, this accout should offer it signer cap to DAO account
+        /// The voting duration in secs
+        voting_duration: u64,
+        /// Total number of eligible member NFTs
+        total_supply: u64,
+        /// Minimum weight for proposal
+        min_proposal_weight: u64,
+        /// Proposal counter
+        cur_largest_proposal_id: u64,
+    }
+
+    struct ResolveThreshold has store {
+        numerator: u64,
+        denominator: u64,
+    }
+
+    /// The collection should only contains NFTs, where all token name only has 1 maximal and immutable
+    /// The total supply is fixed with the token names.
+    struct GovernanceToken has store {
+        /// The creator address of this NFT collection
+        creator: address,
+        /// The collection name
+        collection: String,
+        /// Eligible token_ids that can vote. Value of the table is the weights of the token
+        member_tokens: Table<TokenId, u64>
+    }
+
+    /// All proposals
+    struct Proposals has key {
+        proposals: Table<u64, Proposal>,
+    }
+
+    /// Store the general information about a proposal
+    struct Proposal has copy, drop, store {
+        /// Name of the proposal, limiting to 64 chars
+        name: String,
+        /// Description of the proposal, limiting to 512 chars
+        description: String,
+        /// The name of function to be executed
+        function_name: String,
+        /// The function arguments to be exectued
+        function_args: PropertyMap,
+        /// The start time of the voting
+        start_time_sec: u64,
+        /// Proposal results, unresolved, passed, not passed
+        resolution: u8,
+    }
+
+    struct ProposalVotingStatistics has key {
+        proposals: Table<u64, VotingStatistics>,
+    }
+
+    struct VotingStatistics has store {
+        /// Total yes votes
+        total_yes: u64,
+        /// Total no notes
+        total_no: u64,
+        /// Token voted yes
+        yes_votes: BucketTable<TokenId, address>, // address is the original voter's address for keeping a record of who voted
+        /// Token voted no
+        no_votes: BucketTable<TokenId, address>,
+    }
+
+    struct RegisteredAccounts has key {
```

Comment:
Why is this global instead of created for each DAO? Having one copy per DAO would avoid having to track the admin account separately here.

Reply:
rename it to RegisteredDAOs to make it more clear. This struct stores all the DAO and their signer_capability.  If we store signer capability of each DAO to its resource accounts, the platform won't be able to get the signer to transfer fund on the DAO's behalf. We also don't deploy code to each DAO so there is not a way to acquire signer without storing the signer-capability on the platform contract

Reply:
Isn't it also easy to query whether a DAO exists by checking the existence of the DAO resource? What do you mean by DAO delegation?

In general, I think it makes sense to have this information in the DAO resource as:
1. It's just fewer resources to track and understand (few account locations where they're stored as well).
2. Parallelism. One of the advantages of Aptos Move resources is they can be distributed across accounts, which helps with parallel execution. Having a global record of things when not absolutely would lead to a loss of than advantage.

Reply:
The platform needs to maintain a record which address is the owner of DAO resource account. Later, it would allow these admins to switch to a different admin address. 
I think this mapping data should be stored and maintained by the platform

In general, the parallel is good. But is it a constraint in DAO creation in particular? For an analogy, NFT with 6k minting in a couple of hrs can go through without noticable delay. I would see DAO creation traffic would be less than popular NFT collection minting.

Reply:
I think there are 2 separate fields in RegisteredAccounts, so let me discuss them separately:
1. The DAO account's SignerCap can be part of the DAO resource stored on the DAO account itself right? At least right now, there are no use cases for it, and if there are in the future, the extraction and usage of the SignerCap can be extracted directly by the module itself from the DAO account's DAO resource. What are the benefits of maintaining SignerCap separately in RegisteredAccounts then?
2. The mapping from admin to DAO address could be useful but same question here. What are the benefits of storing this in RegisteredAccounts vs two alternatives:
a. Just emit an event in the DAO resource account during creation with the admin address and when the admin is changed. Indexer can then use the events to construct the mapping.
b. Store the DAO address in a resource on the admin account.

Reply:
sounds good. 
Actually, the signer_capability and admin address can both be stored in the DAO config, which is also stored in the dao resource account.

Reply:
Another related point I just realized: having the global mapping from admin -> DAO address means each account can only administrate one DAO. Might not be too big of a deal but it's a restriction.